cmds/db-manager/main.go

```diff
@@ -0,0 +1,219 @@
+		port:    flag.Int("cockroach_port", 26257, "cockroach port to connect to"),
+		sslMode: flag.String("cockroach_ssl_mode", "disable", "cockroach sslmode"),
+		user:    flag.String("cockroach_user", "root", "cockroach user to authenticate as"),
+		sslDir:  flag.String("cockroach_ssl_dir", "", "directory to ssl certificates. Must contain files: ca.crt, client.<user>.crt, client.<user>.key"),
+	}
+)
+
+func main() {
+	flag.Parse()
+	if *path == "" {
+		log.Panic("Must specify migration_dir path")
+	}
+	if (*dbVersion == "" && *step == 0) || (*dbVersion != "" && *step != 0) {
+		log.Panic("Must specify a db_version or migration_step to goto")
+	}
+	uriParams := map[string]string{
+		"host":             *cockroachParams.host,
+		"port":             strconv.Itoa(*cockroachParams.port),
+		"user":             *cockroachParams.user,
+		"ssl_mode":         *cockroachParams.sslMode,
+		"ssl_dir":          *cockroachParams.sslDir,
+		"application_name": "SchemaManager",
+	}
+	postgresURI, err := dssCockroach.BuildURI(uriParams)
+	if err != nil {
+		log.Panic("Failed to build URI", zap.Error(err))
+	}
+	crdbURI := strings.Replace(postgresURI, "postgresql", "cockroachdb", 1)
+	*path = fmt.Sprintf("file://%v", *path)
+	migrater, migraterErr := migrate.New(*path, crdbURI)
+	myMigrater := &MyMigrate{migrater}
+	defer func() {
+		if migraterErr == nil {
+			if _, err := myMigrater.Close(); err != nil {
+				log.Println(err)
+			}
+		}
+	}()
+	if migraterErr == nil {
+		myMigrater.PrefetchMigrations = 10
+		myMigrater.LockTimeout = time.Duration(15) * time.Second
+
+		// handle Ctrl+c
+		signals := make(chan os.Signal, 1)
+		signal.Notify(signals, syscall.SIGINT)
+		go func() {
+			for range signals {
+				log.Println("Stopping after this running migration ...")
+				myMigrater.GracefulStop <- true
+				return
+			}
+		}()
+	}
+	totalMove := 0
+	migrateDirection := myMigrater.MigrationDirection(postgresURI, *dbVersion, *step)
+	for migrateDirection != 0 {
+		migraterErr = myMigrater.Steps(int(migrateDirection))
+		if migraterErr != nil {
+			log.Fatal(migraterErr)
+		}
+		totalMove += int(migrateDirection)
+		log.Printf("Migrated %s by %d step", migrateDirection.String(), int(math.Abs(float64(int(migrateDirection)))))
+		migrateDirection = myMigrater.MigrationDirection(postgresURI, *dbVersion, *step)
+	}
+	if totalMove == 0 {
+		log.Println("No Changes")
+	} else {
+		log.Printf("Moved %d steps %s in total", int(math.Abs(float64(totalMove))), migrateDirection.String())
+	}
+
+	currentDBVersion, err := getCurrentDBVersion(postgresURI)
+	if err != nil {
+		log.Fatal("Failed to get Current DB version for confirmation")
+	}
+	migrationStep, dirty, err := myMigrater.Version()
+	if err != nil {
+		log.Fatal("Failed to get Migration Step for confirmation")
+	}
+	log.Printf("DB Version: %s, Migration Step # %d, Dirty: %v", currentDBVersion, migrationStep, dirty)
+}
+
+func getCurrentDBVersion(crdbURI string) (string, error) {
```

Comment:
Let's mirror [the migrater's approach](https://github.com/golang-migrate/migrate/blob/8b6e3f1a2ea1fd3016ad479965c573d52469fe51/database/cockroachdb/cockroachdb.go#L256) here since they probably have best practices.  After making the change mentioned above to [construct a migration driver](https://github.com/golang-migrate/migrate/blob/8b6e3f1a2ea1fd3016ad479965c573d52469fe51/database/cockroachdb/cockroachdb.go#L55), we'll have a DB ready to use without redialing.

Reply:
Per the migrater's approach, we can make just one query and check for 42P01 to detect a missing table; can we skip the entire first half of this method with the following?
```go
  const defaultVersion = "v0.0.0"
	err = c.db.QueryRow(getVerisonQuery).Scan(&version)

	switch {
	case err == sql.ErrNoRows:
		return defaultVersion, nil

	case err != nil:
		if e, ok := err.(*pq.Error); ok {
			// 42P01 is "UndefinedTableError" in CockroachDB
			// https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/pgwire/pgerror/codes.go
			if e.Code == "42P01" {
				return defaultVersion, nil
			}
		}
		return "", err

	default:
		return version, nil
	}
```